File: cmps_2016_rf_model.py
"""
CMPS 2016 Random Forest Model
Latino Trump Vote Prediction
"""

import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    confusion_matrix, classification_report, roc_auc_score,
    roc_curve, accuracy_score, precision_score, recall_score, f1_score
)
import shap
import warnings
warnings.filterwarnings('ignore')

print("=" * 70)
print("CMPS 2016 RANDOM FOREST MODEL")
print("Latino Trump Vote Prediction")
print("=" * 70)
print()

# ============================================================
# LOAD PREPARED DATA (from R pipeline)
# ============================================================
print("LOADING PREPARED DATA")
print("-" * 40)

# Load the R data using pyreadr
import pyreadr
result = pyreadr.read_r('cmps_2016_rf_prepared_data.rda')

# Get train and test data
train_data = result['train_data']
test_data = result['test_data']

print(f"Training set: {train_data.shape}")
print(f"Test set: {test_data.shape}")
print(f"Trump in training: {train_data['trump_vote'].sum()}")
print(f"Trump in test: {test_data['trump_vote'].sum()}")
print()

# ============================================================
# PREPARE DATA FOR RF
# ============================================================
print("PREPARING DATA FOR RF")
print("-" * 40)

# Separate features and target
X_train = train_data.drop('trump_vote', axis=1)
y_train = train_data['trump_vote']

X_test = test_data.drop('trump_vote', axis=1)
y_test = test_data['trump_vote']

print(f"Features: {X_train.shape[1]}")
print(f"Training samples: {X_train.shape[0]}")
print(f"Test samples: {X_test.shape[0]}")
print()

# ============================================================
# FIT RANDOM FOREST WITH CLASS WEIGHTS
# ============================================================
print("FITTING RANDOM FOREST")
print("-" * 40)

# Calculate class weights
n_other = (y_train == 0).sum()
n_trump = (y_train == 1).sum()
weight_trump = n_other / n_trump

print(f"Class counts: Other={n_other}, Trump={n_trump}")
print(f"Class weight for Trump: {weight_trump:.3f}")
print()

print("Training Random Forest (this may take a few minutes)...")

rf_model = RandomForestClassifier(
    n_estimators=500,
    max_features='sqrt',
    class_weight='balanced',
    random_state=42,
    n_jobs=-1,
    verbose=1
)

rf_model.fit(X_train, y_train)

print("\nModel trained successfully!")
print()

# ============================================================
# PREDICTIONS
# ============================================================
print("PREDICTIONS")
print("-" * 40)

pred_class = rf_model.predict(X_test)
pred_prob = rf_model.predict_proba(X_test)[:, 1]

print(f"Predictions generated for {len(pred_class)} test samples")
print()

# ============================================================
# EVALUATION: CONFUSION MATRIX
# ============================================================
print("CONFUSION MATRIX")
print("-" * 40)

cm = confusion_matrix(y_test, pred_class)
print("             Predicted")
print("            Other  Trump")
print(f"Actual Other  {cm[0,0]:5d}  {cm[0,1]:5d}")
print(f"       Trump  {cm[1,0]:5d}  {cm[1,1]:5d}")
print()

# Calculate metrics
accuracy = accuracy_score(y_test, pred_class)
precision = precision_score(y_test, pred_class)
recall = recall_score(y_test, pred_class)
f1 = f1_score(y_test, pred_class)
specificity = cm[0, 0] / (cm[0, 0] + cm[0, 1])

print("Metrics:")
print(f"  Accuracy:    {accuracy:.4f}")
print(f"  Precision:   {precision:.4f}")
print(f"  Recall:      {recall:.4f} (Trump correctly identified)")
print(f"  Specificity: {specificity:.4f}")
print(f"  F1 Score:    {f1:.4f}")
print()

# ============================================================
# EVALUATION: AUC-ROC
# ============================================================
print("AUC-ROC")
print("-" * 40)

auc_value = roc_auc_score(y_test, pred_prob)
print(f"AUC: {auc_value:.4f}")

# Save ROC curve data
fpr, tpr, thresholds = roc_curve(y_test, pred_prob)
roc_df = pd.DataFrame({
    'fpr': fpr,
    'tpr': tpr,
    'threshold': thresholds
})
roc_df.to_csv('cmps_2016_rf_roc_curve.csv', index=False)
print("ROC curve data saved to: cmps_2016_rf_roc_curve.csv")
print()

# ============================================================
# VARIABLE IMPORTANCE
# ============================================================
print("VARIABLE IMPORTANCE")
print("-" * 40)

importance_df = pd.DataFrame({
    'variable': X_train.columns,
    'importance': rf_model.feature_importances_
}).sort_values('importance', ascending=False)

print("Top 30 variables by importance:")
print("-" * 70)
print(f"{'Variable':<55} {'Importance':>12}")
print("-" * 70)

for i, row in importance_df.head(30).iterrows():
    print(f"{row['variable'][:55]:<55} {row['importance']:>12.6f}")

# Save full importance
importance_df.to_csv('cmps_2016_rf_importance.csv', index=False)
print(f"\nFull importance saved to: cmps_2016_rf_importance.csv")
print()

# ============================================================
# SHAP VALUES
# ============================================================
print("SHAP VALUES")
print("-" * 40)

print("Computing SHAP values (this may take a while)...")

try:
    # Use a smaller sample for SHAP to speed up computation
    shap_sample_size = min(500, len(X_test))
    X_shap = X_test.iloc[:shap_sample_size]

    # Create SHAP explainer
    explainer = shap.TreeExplainer(rf_model)
    shap_values = explainer.shap_values(X_shap)

    # For binary classification, shap_values is a list [class_0, class_1]
    # We want class 1 (Trump)
    if isinstance(shap_values, list):
        shap_values_trump = shap_values[1]
    else:
        # New SHAP versions return 3D array for multi-output
        if len(shap_values.shape) == 3:
            shap_values_trump = shap_values[:, :, 1]
        else:
            shap_values_trump = shap_values

    # Ensure 2D array
    if len(shap_values_trump.shape) == 1:
        shap_values_trump = shap_values_trump.reshape(-1, 1)

    # Calculate mean absolute SHAP values
    mean_shap = np.abs(shap_values_trump).mean(axis=0)

    # Create dataframe
    shap_importance = pd.DataFrame({
        'variable': X_train.columns.tolist(),
        'mean_abs_shap': mean_shap.flatten()
    }).sort_values('mean_abs_shap', ascending=False)

    print("\nTop 30 variables by mean |SHAP|:")
    print("-" * 70)
    print(f"{'Variable':<55} {'Mean |SHAP|':>12}")
    print("-" * 70)

    for _, row in shap_importance.head(30).iterrows():
        print(f"{row['variable'][:55]:<55} {row['mean_abs_shap']:>12.6f}")

    # Save SHAP importance
    shap_importance.to_csv('cmps_2016_rf_shap_importance.csv', index=False)
    print(f"\nSHAP importance saved to: cmps_2016_rf_shap_importance.csv")

    # Save raw SHAP values for the sample
    shap_df = pd.DataFrame(shap_values_trump, columns=X_train.columns)
    shap_df.to_csv('cmps_2016_rf_shap_values.csv', index=False)
    print(f"Raw SHAP values saved to: cmps_2016_rf_shap_values.csv")

except Exception as e:
    print(f"SHAP computation failed: {e}")
    print("Skipping SHAP values, using feature importance instead.")

print()

# ============================================================
# SAVE PREDICTIONS
# ============================================================
print("SAVING OUTPUTS")
print("-" * 40)

predictions_df = pd.DataFrame({
    'actual': y_test.values,
    'predicted': pred_class,
    'prob_trump': pred_prob
})
predictions_df.to_csv('cmps_2016_rf_predictions.csv', index=False)
print("Predictions saved to: cmps_2016_rf_predictions.csv")

# Save model (using joblib)
import joblib
joblib.dump(rf_model, 'cmps_2016_rf_model.joblib')
print("Model saved to: cmps_2016_rf_model.joblib")
print()

# ============================================================
# FINAL SUMMARY
# ============================================================
print()
print("=" * 70)
print("MODEL SUMMARY")
print("=" * 70)
print(f"Algorithm:      Random Forest")
print(f"Trees:          500")
print(f"Features:       {X_train.shape[1]}")
print(f"Class weights:  balanced")
print()
print("Test Set Performance:")
print(f"  AUC:          {auc_value:.4f}")
print(f"  Accuracy:     {accuracy:.4f}")
print(f"  Precision:    {precision:.4f}")
print(f"  Recall:       {recall:.4f}")
print(f"  F1:           {f1:.4f}")
print("=" * 70)

File: cmps_2016_rf_model.R
# ============================================================
# CMPS 2016 Random Forest Model
# Latino Trump Vote Prediction
# ============================================================

cat(strrep("=", 70), "\n")
cat("CMPS 2016 RANDOM FOREST MODEL\n")
cat("Latino Trump Vote Prediction\n")
cat(strrep("=", 70), "\n\n")

# Check for required packages
if (!require("randomForest", quietly = TRUE)) {
  cat("Installing randomForest package...\n")
  install.packages("randomForest", repos = "https://cloud.r-project.org")
  library(randomForest)
}

if (!require("pROC", quietly = TRUE)) {
  cat("Installing pROC package...\n")
  install.packages("pROC", repos = "https://cloud.r-project.org")
  library(pROC)
}

# ============================================================
# LOAD PREPARED DATA
# ============================================================
cat("LOADING PREPARED DATA\n")
cat(strrep("-", 40), "\n")

load("cmps_2016_rf_prepared_data.rda")
cat("Training set:", nrow(train_data), "x", ncol(train_data), "\n")
cat("Test set:", nrow(test_data), "x", ncol(test_data), "\n")
cat("Trump in training:", sum(train_data$trump_vote), "\n")
cat("Trump in test:", sum(test_data$trump_vote), "\n\n")

# ============================================================
# PREPARE DATA FOR RF
# ============================================================
cat("PREPARING DATA FOR RF\n")
cat(strrep("-", 40), "\n")

# Separate features and target
X_train <- train_data[, setdiff(names(train_data), "trump_vote")]
y_train <- as.factor(train_data$trump_vote)

X_test <- test_data[, setdiff(names(test_data), "trump_vote")]
y_test <- as.factor(test_data$trump_vote)

cat("Features:", ncol(X_train), "\n")
cat("Training samples:", nrow(X_train), "\n")
cat("Test samples:", nrow(X_test), "\n\n")

# ============================================================
# FIT RANDOM FOREST WITH CLASS WEIGHTS
# ============================================================
cat("FITTING RANDOM FOREST\n")
cat(strrep("-", 40), "\n")

# Calculate class weights (inverse of class frequency)
class_counts <- table(y_train)
class_weights <- max(class_counts) / class_counts
cat("Class weights:\n")
cat("  Class 0 (Other):", round(class_weights["0"], 3), "\n")
cat("  Class 1 (Trump):", round(class_weights["1"], 3), "\n\n")

# Fit RF with stratified sampling (sampsize) to handle imbalance
# Using classwt for weighting
set.seed(42)
cat("Training Random Forest (this may take a few minutes)...\n")

rf_model <- randomForest(
  x = X_train,
  y = y_train,
  ntree = 500,
  mtry = floor(sqrt(ncol(X_train))),  # sqrt(p) for classification
  classwt = c("0" = 1, "1" = class_weights["1"]),  # Weight minority class
  importance = TRUE,
  do.trace = 100
)

cat("\nModel trained successfully!\n")
print(rf_model)

# ============================================================
# PREDICTIONS
# ============================================================
cat("\n\nPREDICTIONS\n")
cat(strrep("-", 40), "\n")

# Predict on test set
pred_class <- predict(rf_model, X_test, type = "response")
pred_prob <- predict(rf_model, X_test, type = "prob")[, "1"]

cat("Predictions generated for", length(pred_class), "test samples\n\n")

# ============================================================
# EVALUATION: CONFUSION MATRIX
# ============================================================
cat("CONFUSION MATRIX\n")
cat(strrep("-", 40), "\n")

conf_matrix <- table(Predicted = pred_class, Actual = y_test)
print(conf_matrix)

# Calculate metrics
TP <- conf_matrix["1", "1"]
TN <- conf_matrix["0", "0"]
FP <- conf_matrix["1", "0"]
FN <- conf_matrix["0", "1"]

accuracy <- (TP + TN) / sum(conf_matrix)
precision <- TP / (TP + FP)
recall <- TP / (TP + FN)  # Sensitivity
specificity <- TN / (TN + FP)
f1 <- 2 * precision * recall / (precision + recall)

cat("\nMetrics:\n")
cat("  Accuracy:    ", round(accuracy, 4), "\n")
cat("  Precision:   ", round(precision, 4), "\n")
cat("  Recall:      ", round(recall, 4), "(Trump correctly identified)\n")
cat("  Specificity: ", round(specificity, 4), "\n")
cat("  F1 Score:    ", round(f1, 4), "\n")

# ============================================================
# EVALUATION: AUC-ROC
# ============================================================
cat("\n\nAUC-ROC\n")
cat(strrep("-", 40), "\n")

roc_obj <- roc(y_test, pred_prob, levels = c("0", "1"), direction = "<")
auc_value <- auc(roc_obj)
cat("AUC:", round(auc_value, 4), "\n")

# Save ROC curve data
roc_data <- data.frame(
  specificity = roc_obj$specificities,
  sensitivity = roc_obj$sensitivities
)
write.csv(roc_data, "cmps_2016_rf_roc_curve.csv", row.names = FALSE)
cat("ROC curve data saved to: cmps_2016_rf_roc_curve.csv\n")

# ============================================================
# VARIABLE IMPORTANCE
# ============================================================
cat("\n\nVARIABLE IMPORTANCE\n")
cat(strrep("-", 40), "\n")

# Get importance measures
importance_df <- data.frame(
  variable = rownames(rf_model$importance),
  MeanDecreaseAccuracy = rf_model$importance[, "MeanDecreaseAccuracy"],
  MeanDecreaseGini = rf_model$importance[, "MeanDecreaseGini"],
  stringsAsFactors = FALSE
)

# Sort by MeanDecreaseAccuracy
importance_df <- importance_df[order(-importance_df$MeanDecreaseAccuracy), ]

cat("Top 30 variables by MeanDecreaseAccuracy:\n")
cat(strrep("-", 70), "\n")
cat(sprintf("%-50s %15s %15s\n", "Variable", "DecAccuracy", "DecGini"))
cat(strrep("-", 70), "\n")

for (i in 1:min(30, nrow(importance_df))) {
  cat(sprintf("%-50s %15.4f %15.4f\n",
              substr(importance_df$variable[i], 1, 50),
              importance_df$MeanDecreaseAccuracy[i],
              importance_df$MeanDecreaseGini[i]))
}

# Save full importance
write.csv(importance_df, "cmps_2016_rf_importance.csv", row.names = FALSE)
cat("\nFull importance saved to: cmps_2016_rf_importance.csv\n")

# ============================================================
# SAVE MODEL
# ============================================================
cat("\n\nSAVING MODEL\n")
cat(strrep("-", 40), "\n")

save(rf_model, file = "cmps_2016_rf_model.rda")
cat("Model saved to: cmps_2016_rf_model.rda\n")

# Save predictions
predictions_df <- data.frame(
  actual = y_test,
  predicted = pred_class,
  prob_trump = pred_prob
)
write.csv(predictions_df, "cmps_2016_rf_predictions.csv", row.names = FALSE)
cat("Predictions saved to: cmps_2016_rf_predictions.csv\n")

# ============================================================
# FINAL SUMMARY
# ============================================================
cat("\n\n")
cat(strrep("=", 70), "\n")
cat("MODEL SUMMARY\n")
cat(strrep("=", 70), "\n")
cat("Algorithm:      Random Forest\n")
cat("Trees:          500\n")
cat("Features:      ", ncol(X_train), "\n")
cat("Class weights:  1.0 (Other), ", round(class_weights["1"], 2), " (Trump)\n")
cat("\n")
cat("Test Set Performance:\n")
cat("  AUC:          ", round(auc_value, 4), "\n")
cat("  Accuracy:     ", round(accuracy, 4), "\n")
cat("  Precision:    ", round(precision, 4), "\n")
cat("  Recall:       ", round(recall, 4), "\n")
cat("  F1:           ", round(f1, 4), "\n")
cat(strrep("=", 70), "\n")
